<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <title>Metagenomics by camilo-v</title>

    <link rel="stylesheet" href="stylesheets/styles.css">
    <link rel="stylesheet" href="stylesheets/pygment_trac.css">
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="wrapper">
      <header>
        <h1>Metagenomics</h1>
        <p>Software related to Metagenomics, Bioinformatics, and Statistical methods &amp; publications.</p>

        <p class="view"><a href="https://github.com/camilo-v/Metagenomics">View the Project on GitHub <small>camilo-v/Metagenomics</small></a></p>


        <ul>
          <li><a href="https://github.com/camilo-v/Metagenomics/zipball/master">Download <strong>ZIP File</strong></a></li>
          <li><a href="https://github.com/camilo-v/Metagenomics/tarball/master">Download <strong>TAR Ball</strong></a></li>
          <li><a href="https://github.com/camilo-v/Metagenomics">View On <strong>GitHub</strong></a></li>
        </ul>
      </header>
      <section>
        <h3>
<a id="detecting-bacterial-genomes-in-a-metagenomic-sample-using-ngs-reads" class="anchor" href="#detecting-bacterial-genomes-in-a-metagenomic-sample-using-ngs-reads" aria-hidden="true"><span class="octicon octicon-link"></span></a>Detecting Bacterial Genomes in a Metagenomic Sample Using NGS Reads.</h3>

<p>The paper proposes a system for detecting the presence of strains of bacteria in metagenomic samples generated using whole-genome hight-throughput DNA sequencing, or Next-Generation Sequencing (NGS).</p>

<p>We use a nucleotide flipping technique on whole genome next generation sequencing (NGS) data to test for the pres- ence of various bacterial strains in a single metagenomic sample.  We introduce artificial point mutations on the NGS reads at the nucleotide level to define a test statistic for each genome in bacterial reference database; we also use a variant of the Westfall-Young procedure to correct for multiple comparisons.</p>

<p>You can obtain a copy of the source code by:</p>

<pre><code>$ cd your_computer/your_repo_root/repo_name
$ git clone https://github.com/camilo-v/Metagenomics.git
</code></pre>

<p>If you're using the GitHub for Mac, simply sync your repository and you'll see the new branch.</p>

<h3>
<a id="requirements" class="anchor" href="#requirements" aria-hidden="true"><span class="octicon octicon-link"></span></a>Requirements</h3>

<p>The software was developed on OS X (10.9) and covers four (4) programming languages:</p>

<ul>
<li>Python (2.7 variant)</li>
<li>Perl (perl 5, v.16 or above)</li>
<li>R (3.1 or above)</li>
<li>Bash Shell</li>
</ul>

<p>You will need a reasonably modern computer and operating system that can support all four.</p>

<h4>
<a id="bowtie2" class="anchor" href="#bowtie2" aria-hidden="true"><span class="octicon octicon-link"></span></a>Bowtie2</h4>

<p><a href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml">Bowtie2</a> is required for the alignment step.  See the Bowtie2 <a href="http://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#getting-started-with-bowtie-2-lambda-phage-example">Getting Started</a> guide for instructions on installing and running Bowtie2.</p>

<h4>
<a id="samtools" class="anchor" href="#samtools" aria-hidden="true"><span class="octicon octicon-link"></span></a>Samtools</h4>

<p><a href="http://www.htslib.org/">Samtools</a> is required for parsing the alignments and generating miscellaneous alignment statistics.  See the <a href="http://www.htslib.org/doc/">Samtools Documentation</a> for instructions on installing and running Samtools.</p>

<h3>
<a id="maintainers" class="anchor" href="#maintainers" aria-hidden="true"><span class="octicon octicon-link"></span></a>Maintainers</h3>

<ul>
<li><a href="mailto:cvaldes3@miami.edu">Camilo Valdes</a></li>
</ul>

<h3>
<a id="collaborators" class="anchor" href="#collaborators" aria-hidden="true"><span class="octicon octicon-link"></span></a>Collaborators</h3>

<ul>
<li><a href="mailto:jclarke3@unl.edu">Jennifer Clarke</a></li>
<li><a href="mailto:bclarke3@unl.edu">Bertrand Clarke</a></li>
</ul>

<h2>
<a id="contact" class="anchor" href="#contact" aria-hidden="true"><span class="octicon octicon-link"></span></a>Contact</h2>

<p>Contact <a href="mailto:cvaldes3@miami.edu">Camilo Valdes</a> for pull requests, and/or bug reports.</p>
      </section>
      <footer>
        <p>This project is maintained by <a href="https://github.com/camilo-v">camilo-v</a></p>
        <p><small>Hosted on GitHub Pages &mdash; Theme by <a href="https://github.com/orderedlist">orderedlist</a></small></p>
      </footer>
    </div>
    <script src="javascripts/scale.fix.js"></script>
    
  </body>
</html>